Documentation v2

sankethkatta opened the pull request:
+ Reorganize and rewrite README for clarity. See it here:
  - [documentation-v2/README.md](https://github.com/smartcar/javascript-sdk/blob/documentation-v2/README.md)
+ Autogenerate reference documentation with `jsdoc-to-markdown`. See in:
  - [documentation-v2/doc/README.md](https://github.com/smartcar/javascript-sdk/blob/documentation-v2/doc/README.md)
+ Template package version number into README using `gulp-template`. All changes to README should now be made in `README.mdt` and templated using `npm run readme`.
+ Method references are linked from README into the autogenerated JSDoc.

sankethkatta commented on README.mdt:
```diff
@@ -0,0 +1,177 @@
+
+### npm
+
+**TODO:** Add package name when publishing to npm
```
@gurpreetatwal you'll need to edit this when publishing to npm

gurpreetatwal commented on README.mdt:
```diff
@@ -0,0 +1,177 @@
+
+To make requests to a vehicle from a web application, the end user must connect their vehicle using [Smartcar's authorization flow](https://smartcar.com/docs#authentication). The Smartcar JavaScript SDK provides an easy way to launch and handle the authorization flow to retrieve the resulting `code`.
+
+Before integrating with the JavaScript SDK, you'll need to register an application in the [Smartcar dashboard](https://dashboard.smartcar.com). Once you are registered, you will have a Client ID and Client Secret which will allow you to authorize users.
```
this last sentence sounds kind of odd to me, but deferring to @charlottekosche here

sankethkatta replied:
What about the inverting the first sentence:

>  Register an application in the [Smartcar dashboard](https://dashboard.smartcar.com) before integrating with the JavaScript SDK. Once you are registered, you will have a Client ID and Client Secret which will allow you to authorize users.

esonmez replied:
I defer to @charlottekosche but the current version sounds better to me than the inverted version.

sankethkatta commented on src/sdk.js:
```diff
@@ -207,10 +210,10 @@ Smartcar._getWindowOptions = function() {
  * screen to show on every authentication, even if the user has previously
  * consented to the exact scope of permission
- * @return {String} - generated OAuth link
+ * @return {String} generated OAuth link
```
:+1: I deleted the return line as this is a void. Then I moved `return false` into the `addEventListener` callback.

charlottekosche commented on README.mdt:
```diff
@@ -0,0 +1,177 @@
+
+The SDK facilitates generating OAuth links, creating pop-up dialogs, and recieving authorization codes. This SDK will not assist in exchanging authorization codes for an access token or making requests to vehicles. Please see our [Backend SDKs](https://smartcar.com/docs#backend-sdks) for more on handling the access tokens and vehicle requests.
+
+## Quick Start
```
## Quick **start**

charlottekosche commented on README.mdt:
```diff
@@ -0,0 +1,177 @@
+
+## Quick Start
+
+### 1. Register a JavaScript SDK Redirect URI
```
**redirect** URI

charlottekosche commented on README.mdt:
```diff
@@ -0,0 +1,177 @@
+
+**Reference:** [`new Smartcar(options)`](doc#new_Smartcar_new)
+
+> NOTE: See the full set of available scopes in the [Smartcar API Reference](https://smartcar.com/docs#get-all-vehicles) under "Required Permissions" for each endpoint.
```
> NOTE: See the full set of available scopes **for each endpoint** in the [Smartcar API **reference**] (https://smartcar.com/docs#get-all-vehicles) under "Required **Permissions."**

charlottekosche commented on README.mdt:
```diff
@@ -0,0 +1,177 @@
+
+## Advanced
+
+The JavaScript SDK also can be used in alternative flows than the one described above. This section will cover some of its other usages.
```
**In addition to the flow described above, you can use the JavaScript SDK in other ways too. The following** section will cover some of **these cases**.

charlottekosche commented on README.mdt:
```diff
@@ -0,0 +1,177 @@
+
+The JavaScript SDK also can be used in alternative flows than the one described above. This section will cover some of its other usages.
+
+### Authorization URI Generation
```
**generation**

sankethkatta replied:
I also changed it to "Authorization **URL** Generation" as the function is called `getAuthUrl()`. This is different than the redirect URI.

charlottekosche commented on README.mdt:
```diff
@@ -0,0 +1,177 @@
+
+### Authorization URI Generation
+
+Normally the [`.addClickHandler()`](doc#Smartcar+addClickHandler) and [`.openDialog()`](doc#Smartcar+openDialog) methods are used to launch the authorization flow. However, if you would like to generate the OAuth authorization URL directly you can do so with the [`.getAuthUrl()`](doc#Smartcar+getAuthUrl) method.
```
However, if you would like to generate the OAuth authorization URL **directly,**

charlottekosche commented on README.mdt:
```diff
@@ -0,0 +1,177 @@
+
+**Reference:** [`smartcar.getAuthUrl(options)`](doc#Smartcar+getAuthUrl)
+
+### Server side redirect handling
```
**Server-side**

charlottekosche commented on README.mdt:
```diff
@@ -0,0 +1,177 @@
+
+### Server side redirect handling
+
+In a traditional OAuth implementation, the redirect uri is normally set to your application's server backend, rather than the JavaScript SDK redirect page used by this SDK. In this architecture you would recieve the authorization code on a backend route instead of the client-side `onComplete` callback.
```
In a traditional OAuth implementation, the redirect **URI** is normally set to your application's server **back end**, rather than **to a** redirect page used by **an SDK**. In **the architecture of this JavaScript SDK**, you **receive** the authorization code on a **back-end** route instead of the client-side `onComplete` callback.

charlottekosche replied:
I'm not sure whether I understand these two sentences correctly, so my edits might be wrong.

charlottekosche commented on README.mdt:
```diff
@@ -0,0 +1,177 @@
+
+In a traditional OAuth implementation, the redirect uri is normally set to your application's server backend, rather than the JavaScript SDK redirect page used by this SDK. In this architecture you would recieve the authorization code on a backend route instead of the client-side `onComplete` callback.
+
+To use the JavaScript SDK for this flow:
```
To use the JavaScript SDK for this **flow, do the following**:

charlottekosche commented on README.mdt:
```diff
@@ -0,0 +1,177 @@
+
+To use the JavaScript SDK for this flow:
+
+1. Set the `redirect_uri` parameter in the initialization to a route on your application's backend server:
```
**back-end** server

charlottekosche commented on README.mdt:
````diff
@@ -0,0 +1,177 @@
+});
+```
+
+Make sure to also register the uri on the [Smartcar dashboard](https://dashboard.smartcar.com).
````
**URI**

charlottekosche commented on README.mdt:
```diff
@@ -0,0 +1,177 @@
+
+Make sure to also register the uri on the [Smartcar dashboard](https://dashboard.smartcar.com).
+
+2. On your `redirect_uri` route, you will need to accept the authorization code according to the query parameters documented in the [Smartcar API Reference](https://smartcar.com/docs#3-handle-smartcar-response).
```
[Smartcar API **reference**]

charlottekosche commented on doc/README.md:
```diff
@@ -0,0 +1,118 @@
+<a name="Smartcar+openDialog"></a>
+
+### smartcar.openDialog(options) ⇒ <code>String</code>
+Launch the OAuth dialog flow
```
Launches

charlottekosche commented on doc/README.md:
```diff
@@ -0,0 +1,118 @@
+<a name="Smartcar+addClickHandler"></a>
+
+### smartcar.addClickHandler(options)
+Add an on-click event listener to the element with the provided id.
```
**Adds** ... provided **id** (no period)

sankethkatta replied:
I kept the period at the end, as this is auto-generated from the inline JSDoc in the code. By convention the first line of the inline comment ends in a period. See: http://usejsdoc.org/

This is where it is generated from: 
https://github.com/smartcar/javascript-sdk/blob/51098a3e5be0d08929755fb83f81310ca2751557/src/sdk.js#L220-L231

charlottekosche commented on doc/README.md:
```diff
@@ -0,0 +1,118 @@
+
+### smartcar.addClickHandler(options)
+Add an on-click event listener to the element with the provided id.
+On-click event calls openDialog when the specified element is clicked.
```
On-click event calls openDialog when the specified element is **clicked** (no period)

charlottekosche commented on doc/README.md:
```diff
@@ -0,0 +1,118 @@
+<a name="Smartcar.AccessDenied"></a>
+
+### Smartcar.AccessDenied ⇐ <code>Error</code>
+Access denied error returned by authorization flow.
```
Access denied error returned by authorization **flow** (no period)

sankethkatta replied:
See comment above regarding JSDoc convention.

sankethkatta commented on README.mdt:
```diff
@@ -0,0 +1,177 @@
+
+1. User clicks "Connect your car" button (or similar button) on your application's frontend.
+2. User sees a pop-up dialog with the Smartcar authorization flow.
+   1. The user will select the make of their vehicle.
```
I changed it to "User **selects** the make...**

sankethkatta commented:
@charlottekosche I addressed your comments in the latest commit. Please look through and let me know if it looks good! The "Server side redirect" section might still need some work.
